### DISCUSSION TITLE
Obtaining Embeddings in PYG.KGE and inverting to recover original input data

### DISCUSSION THREAD

**Discussion by blainehill2001** [pyg-team/pytorch_geometric#9065] [Q&A] [answered]
Hello,

As I understand in pytorch geometric models such as RotatE or TransE we can obtain an embedded triplet for some input triplet once we've trained our model via node_emb or node_emb_im in the case of RotatE. How **can we recover the original triplet** if we have the embedded triplet? If we can't guarantee the reverse process since we are mapping to a latent space with fewer dimensions than the input dimensions, how can we get the most likely triplet that matches our embedded triplet? Looking at the source code for `torch_geometric.nn.kge` there doesn't seem to be a `decode()` or anything like there is in [this example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/link_pred.py#L42).

To ground the question in a more detailed manner:
Let's say I have a trained embedding model and wish to do link prediction or some reasoning task. I take my (head_node_index, unknown_index, tail_node_index) and embed it using my model (let's say RotatE). 

I now have six objects: model.node_emb(head_node_index), model.node_emb_im(head_node_index), model.node_emb(tail_node_index), model.node_emb_im(tail_node_index), model.rel_emb(unknown_index), model.rel_emb_im(unknown_index)

Then I plug these objects as input to some other model which yields an answer for the link prediction in the embedding space. I now wish to map my six objects to the input dimensions so I can have a meaningful answer.

**Comment by rusty1s** [comment, marked as answer]
Sorry for super late reply :( Is there a point in a `decode` function if you already node `head_node_index`, `tail_node_index` and `unknown_index` (since you receive the embedding from it)? Why do you need to recover these indices from the embedding itself? Otherwise, I would assume you can simply search for the index in `node_emb` that matches with `model.node_emb(head_node_index)`?

**Reply by blainehill2001** [reply]
The goal is to obtain a predicted triplet in the input dimensions (aka natural language such as English) from a embedded triplet/sextuplet. 

I think you've answered the question with your last line: if I have some sextuplet, I take each value and search for the index that yields nearest it. I originally wondered if there was any predefined function to do this (like seen in the example link).
